File: openai/sensor_event_loop.cpp
// C++ twin of sensor_event_loop.py — Embedded multi-sensor Event Loop demo.
//
// Design doc: embedded-sensor-event-loop.md
// Python twin: sensor_event_loop.py
//
// Highlights vs Python:
// - Real acquire/release atomics on SPSC head/tail
// - alignas(64) to avoid False Sharing
// - condition_variable models epoll_wait / futex sleep

#include <atomic>
#include <cassert>
#include <chrono>
#include <cmath>
#include <condition_variable>
#include <cstddef>
#include <cstdint>
#include <deque>
#include <iostream>
#include <mutex>
#include <optional>
#include <thread>
#include <utility>
#include <vector>

constexpr std::size_t kCacheLine = 64;

template <typename T, std::size_t Capacity>
class LockFreeSpscQueue {
    static_assert(Capacity >= 2 && (Capacity & (Capacity - 1)) == 0,
                  "Capacity must be power of 2");

 public:
    bool push(const T& item) {
        const std::size_t tail = tail_.load(std::memory_order_relaxed);
        const std::size_t head = head_.load(std::memory_order_acquire);
        // Free-running counters: full when Cap items are in flight.
        if (tail - head == Capacity) {
            return false;
        }
        buffer_[tail & kMask] = item;
        tail_.store(tail + 1, std::memory_order_release);
        return true;
    }

    std::optional<T> pop() {
        const std::size_t head = head_.load(std::memory_order_relaxed);
        const std::size_t tail = tail_.load(std::memory_order_acquire);
        if (head == tail) {
            return std::nullopt;
        }
        T item = buffer_[head & kMask];
        head_.store(head + 1, std::memory_order_release);
        return item;
    }

 private:
    static constexpr std::size_t kMask = Capacity - 1;
    T buffer_[Capacity]{};
    alignas(kCacheLine) std::atomic<std::size_t> head_{0};
    alignas(kCacheLine) std::atomic<std::size_t> tail_{0};
};

enum class SensorType { Imu, Camera, Audio };

struct SensorEvent {
    SensorType type = SensorType::Imu;
    double timestamp_ms = 0.0;
    std::int64_t payload_id = 0;  // stand-in for dma-buf / frame id
};

struct FusedFrame {
    double cam_t_ms = 0.0;
    std::int64_t cam_payload = 0;
    int imu_count = 0;
};

class EventLoop {
 public:
    void dispatch(const SensorEvent& event) {
        bool ok = true;
        switch (event.type) {
            case SensorType::Imu:
                ok = imu_q_.push(event);
                if (!ok) {
                    ++drops_imu_;
                }
                break;
            case SensorType::Camera:
                ok = cam_q_.push(event);
                if (!ok) {
                    ++drops_cam_;
                }
                break;
            case SensorType::Audio:
                ok = audio_q_.push(event);
                if (!ok) {
                    ++drops_audio_;
                }
                break;
        }
        wakeup();
    }

    void stop() {
        running_.store(false, std::memory_order_relaxed);
        wakeup();
    }

    void run() {
        int idle_spins = 0;
        while (running_.load(std::memory_order_relaxed)) {
            bool processed = false;

            // Drain IMU first (highest rate / lowest latency path).
            while (auto ev = imu_q_.pop()) {
                imu_history_.push_back(*ev);
                if (imu_history_.size() > 512) {
                    imu_history_.pop_front();
                }
                processed = true;
            }

            while (auto ev = cam_q_.pop()) {
                fuse_camera(*ev);
                processed = true;
            }

            while (audio_q_.pop()) {
                processed = true;
            }

            if (processed) {
                idle_spins = 0;
                continue;
            }

            ++idle_spins;
            if (idle_spins < 3) {
                std::this_thread::sleep_for(std::chrono::microseconds(500));
            } else {
                // Power-saving sleep — production: epoll_wait / futex.
                std::unique_lock lock(wake_mu_);
                wake_cv_.wait_for(lock, std::chrono::milliseconds(50), [this] {
                    return wake_pending_ ||
                           !running_.load(std::memory_order_relaxed);
                });
                wake_pending_ = false;
                idle_spins = 0;
            }
        }
    }

    const std::vector<FusedFrame>& fused() const { return fused_; }
    int drops_imu() const { return drops_imu_; }
    int drops_cam() const { return drops_cam_; }
    int drops_audio() const { return drops_audio_; }

 private:
    void wakeup() {
        {
            std::lock_guard lock(wake_mu_);
            wake_pending_ = true;
        }
        wake_cv_.notify_one();
    }

    void fuse_camera(const SensorEvent& cam) {
        constexpr double kWindowMs = 2.5;
        int matched = 0;
        for (const auto& imu : imu_history_) {
            if (std::fabs(imu.timestamp_ms - cam.timestamp_ms) <= kWindowMs) {
                ++matched;
            }
        }
        fused_.push_back(
            FusedFrame{cam.timestamp_ms, cam.payload_id, matched});
    }

    std::atomic<bool> running_{true};

    LockFreeSpscQueue<SensorEvent, 256> imu_q_;
    LockFreeSpscQueue<SensorEvent, 32> cam_q_;
    LockFreeSpscQueue<SensorEvent, 64> audio_q_;

    std::mutex wake_mu_;
    std::condition_variable wake_cv_;
    bool wake_pending_ = false;

    std::deque<SensorEvent> imu_history_;
    std::vector<FusedFrame> fused_;
    int drops_imu_ = 0;
    int drops_cam_ = 0;
    int drops_audio_ = 0;
};

static void producer(EventLoop& loop, SensorType type, double hz,
                     double duration_s) {
    using clock = std::chrono::steady_clock;
    const auto period = std::chrono::duration<double>(1.0 / hz);
    const auto t0 = clock::now();
    std::int64_t n = 0;

    while (std::chrono::duration<double>(clock::now() - t0).count() <
           duration_s) {
        const double now_ms =
            std::chrono::duration<double, std::milli>(clock::now() - t0)
                .count();
        loop.dispatch(SensorEvent{type, now_ms, n});
        ++n;
        std::this_thread::sleep_for(period);
    }
}

int main() {
    EventLoop loop;
    std::thread loop_thread([&] { loop.run(); });

    constexpr double kDuration = 0.35;
    std::thread imu([&] { producer(loop, SensorType::Imu, 200.0, kDuration); });
    std::thread cam(
        [&] { producer(loop, SensorType::Camera, 30.0, kDuration); });
    std::thread audio(
        [&] { producer(loop, SensorType::Audio, 50.0, kDuration); });

    imu.join();
    cam.join();
    audio.join();

    std::this_thread::sleep_for(std::chrono::milliseconds(50));
    loop.stop();
    loop_thread.join();

    const auto& fused = loop.fused();
    assert(!fused.empty());

    int with_imu = 0;
    for (const auto& f : fused) {
        if (f.imu_count > 0) {
            ++with_imu;
        }
    }
    assert(with_imu >= static_cast<int>(fused.size()) / 2);

    const auto& sample = fused[fused.size() / 2];
    std::cout << "fused_frames=" << fused.size() << " with_imu=" << with_imu
              << "\n";
    std::cout << "drops={imu:" << loop.drops_imu()
              << ", cam:" << loop.drops_cam()
              << ", audio:" << loop.drops_audio() << "}\n";
    std::cout << "sample: cam_t=" << sample.cam_t_ms
              << " payload=" << sample.cam_payload
              << " imu_count=" << sample.imu_count << "\n";
    std::cout << "sensor_event_loop_cpp: ok\n";
    return 0;
}

File: openai/sensor_event_loop.py
#!/usr/bin/env python3
"""Embedded multi-sensor event loop demo (SPSC rings + poll/sleep + time sync).

Mirrors the OpenAI Embedded Experiences design question:
  - One SPSC queue per sensor (no multi-producer CAS)
  - Single event-loop thread drains queues (Reactor)
  - Busy-poll while hot; block on Event when idle (power saving)
  - On camera frame, align nearby IMU samples by timestamp window
"""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass
from enum import Enum, auto
from typing import Generic, List, Optional, TypeVar

T = TypeVar("T")


class SpscRing(Generic[T]):
    """Power-of-two SPSC ring. One producer thread, one consumer thread."""

    def __init__(self, capacity: int) -> None:
        if capacity < 2 or (capacity & (capacity - 1)) != 0:
            raise ValueError("capacity must be power of 2 and >= 2")
        self._cap = capacity
        self._mask = capacity - 1
        self._buf: List[Optional[T]] = [None] * capacity
        self._head = 0  # consumer
        self._tail = 0  # producer
        # In C++ these are atomics with acquire/release + alignas(64).
        # CPython GIL makes the demo race-free enough for teaching.

    def push(self, item: T) -> bool:
        if ((self._tail + 1) & self._mask) == (self._head & self._mask):
            return False  # full — drop newest (caller's policy)
        self._buf[self._tail & self._mask] = item
        self._tail += 1
        return True

    def pop(self) -> Optional[T]:
        if self._head == self._tail:
            return None
        item = self._buf[self._head & self._mask]
        self._buf[self._head & self._mask] = None
        self._head += 1
        return item


class SensorType(Enum):
    IMU = auto()
    CAMERA = auto()
    AUDIO = auto()


@dataclass
class SensorEvent:
    type: SensorType
    timestamp_ms: float
    payload: object  # IMU vec / frame id / audio chunk id — zero-copy stand-in


class EventLoop:
    def __init__(self) -> None:
        self.imu_q: SpscRing[SensorEvent] = SpscRing(256)
        self.cam_q: SpscRing[SensorEvent] = SpscRing(32)
        self.audio_q: SpscRing[SensorEvent] = SpscRing(64)
        self._wake = threading.Event()
        self._running = True
        self.imu_history: List[SensorEvent] = []
        self.fused: List[dict] = []
        self.drops = {"imu": 0, "cam": 0, "audio": 0}

    def dispatch(self, event: SensorEvent) -> None:
        ok = True
        if event.type is SensorType.IMU:
            ok = self.imu_q.push(event)
            if not ok:
                self.drops["imu"] += 1
        elif event.type is SensorType.CAMERA:
            ok = self.cam_q.push(event)
            if not ok:
                self.drops["cam"] += 1
        else:
            ok = self.audio_q.push(event)
            if not ok:
                self.drops["audio"] += 1
        self._wake.set()  # wake loop if sleeping

    def stop(self) -> None:
        self._running = False
        self._wake.set()

    def run(self) -> None:
        idle_spins = 0
        while self._running:
            processed = False

            # High-priority: drain IMU first (200Hz path).
            while (ev := self.imu_q.pop()) is not None:
                self.imu_history.append(ev)
                if len(self.imu_history) > 512:
                    self.imu_history = self.imu_history[-512:]
                processed = True

            while (ev := self.cam_q.pop()) is not None:
                self._fuse_camera(ev)
                processed = True

            while (ev := self.audio_q.pop()) is not None:
                processed = True  # demo: count only

            if processed:
                idle_spins = 0
                continue

            idle_spins += 1
            if idle_spins < 3:
                # Short active poll window (low wakeup latency).
                time.sleep(0.0005)
            else:
                # Sleep until producer wakes us — power saving.
                self._wake.wait(timeout=0.05)
                self._wake.clear()
                idle_spins = 0

    def _fuse_camera(self, cam: SensorEvent) -> None:
        """Align IMU samples within ±2.5ms of the camera timestamp."""
        window_ms = 2.5
        t = cam.timestamp_ms
        matched = [
            e
            for e in self.imu_history
            if abs(e.timestamp_ms - t) <= window_ms
        ]
        self.fused.append(
            {
                "cam_t": t,
                "cam_payload": cam.payload,
                "imu_count": len(matched),
                "imu_ts": [e.timestamp_ms for e in matched],
            }
        )


def _producer(
    loop: EventLoop,
    kind: SensorType,
    hz: float,
    duration_s: float,
    payload_fn,
) -> None:
    period = 1.0 / hz
    t0 = time.monotonic()
    n = 0
    while time.monotonic() - t0 < duration_s:
        now_ms = (time.monotonic() - t0) * 1000.0
        loop.dispatch(
            SensorEvent(type=kind, timestamp_ms=now_ms, payload=payload_fn(n))
        )
        n += 1
        time.sleep(period)


def main() -> None:
    loop = EventLoop()
    thread = threading.Thread(target=loop.run, name="event-loop", daemon=True)
    thread.start()

    duration = 0.35
    producers = [
        threading.Thread(
            target=_producer,
            args=(loop, SensorType.IMU, 200.0, duration, lambda i: ("gyro", i)),
            daemon=True,
        ),
        threading.Thread(
            target=_producer,
            args=(loop, SensorType.CAMERA, 30.0, duration, lambda i: ("frame", i)),
            daemon=True,
        ),
        threading.Thread(
            target=_producer,
            args=(loop, SensorType.AUDIO, 50.0, duration, lambda i: ("pcm", i)),
            daemon=True,
        ),
    ]
    for p in producers:
        p.start()
    for p in producers:
        p.join()

    time.sleep(0.05)  # drain tail
    loop.stop()
    thread.join(timeout=1.0)

    assert loop.fused, "expected at least one fused camera+imu bundle"
    # Most camera frames should find a nearby IMU sample at 200Hz.
    with_imu = sum(1 for f in loop.fused if f["imu_count"] > 0)
    assert with_imu >= len(loop.fused) // 2

    print(f"fused_frames={len(loop.fused)} with_imu={with_imu}")
    print(f"drops={loop.drops}")
    print("sample:", loop.fused[len(loop.fused) // 2])
    print("sensor_event_loop: ok")


if __name__ == "__main__":
    main()
